File: January/cpp/Leaf Similar Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
     void findLeaf_dfs(TreeNode* nodeNow, std::vector<int>& nodeList) {
        if (!nodeNow) {
            return;
        }

        if (!nodeNow->left && !nodeNow->right) {
            nodeList.push_back(nodeNow->val);
            return;
        }

        findLeaf_dfs(nodeNow->left, nodeList);
        findLeaf_dfs(nodeNow->right, nodeList);
    }
    public:
 bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        std::vector<int> nodeList_root1, nodeList_root2;
        findLeaf_dfs(root1, nodeList_root1);
        findLeaf_dfs(root2, nodeList_root2);
        return nodeList_root1 == nodeList_root2;
    }

   
};

File: January/Java/Leaf-Similar Trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> l=new ArrayList<>();
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
       List<Integer> x=new ArrayList<>();
       int i=0;
       Leafs(root1);
       x.addAll(l);
       l.clear();
       l=Leafs(root2);
       if(l.size()!=x.size())
           return false;  
       while(i<x.size()){
        if(x.get(i)!=l.get(i))
           return false;
        i++;
       }    
       return true;
    }       
    public List<Integer> Leafs(TreeNode root){   
        if(root==null)
           return l;
        Leafs(root.left);
        if(root.left==null&&root.right==null)
              l.add(root.val);    
        Leafs(root.right);
        return l;         
    }
}
